### ISSUE TITLE
Handle DSL-ish lists

### ISSUE THREAD

**Issue by michalmuskala** [WhatsApp/erlfmt#19]
Lists are often used to represent DSLs, one special and common instance of that is iolists.

A concrete example comes from `asn1ct_gen.erl` that combines iolists with atoms to form a formatting DSL. There, for example, formatter converts:

```erl
gen_part_decode_funcs({constructed,bif},TypeName,
		      {_Name,parts,Tag,_Type}) ->
    emit(["  case Data of",nl,
	  "    L when is_list(L) ->",nl,
	  "      'dec_",TypeName,"'(lists:map(fun(X) -> element(1, ",
	  {call,ber,ber_decode_erlang,["X"]},") end, L),",{asis,Tag},");",nl,
	  "    _ ->",nl,
	  "      [Res] = 'dec_",TypeName,"'([Data],",{asis,Tag},"),",nl,
	  "      Res",nl,
	  "  end"]);
```

into

```erl
gen_part_decode_funcs({constructed, bif}, TypeName, {_Name, parts, Tag, _Type}) ->
    emit([
        "  case Data of",
        nl,
        "    L when is_list(L) ->",
        nl,
        "      'dec_",
        TypeName,
        "'(lists:map(fun(X) -> element(1, ",
        {call, ber, ber_decode_erlang, ["X"]},
        ") end, L),",
        {asis, Tag},
        ");",
        nl,
        "    _ ->",
        nl,
        "      [Res] = 'dec_",
        TypeName,
        "'([Data],",
        {asis, Tag},
        "),",
        nl,
        "      Res",
        nl,
        "  end"
    ]);
```

In general, formatting lists works well today, therefore I wouldn't want to see a solution where we don't reformat them at all. At the same time, I recognise that what is produced is strictly worse than the original.

**Comment by KennethL** [comment]
It is not easy for `erlfmt` to know when to preserve the designers new lines and when not to.
- One approach would be to not add or remove any new lines at all
- Another approach would be to treat certain new lines with respect and don't touch them. It must then be possible for the designer to control this *in a simple and natural way*.

In html we have <pre> and in markdown we also have ways of protecting text that should be untouched. What would the corresponding be in Erlang? An attribute in the beginning of the file/module or embedded anywhere. Could also be some trick with comments (ugly, unreliable?)

**Comment by alanz** [comment]
Do we always reformat, or only if the line length is exceeded?

**Comment by michalmuskala** [comment]
Always. `erlfmt` parses everything to AST and re-prints AST from scratch. There's precise location information that allows detecting where line breaks were originally, and some extra metadata about the original source like parenthesis placement. In general, though, it rebuilds the source from scratch.

**Comment by awalterschulze** [comment]
We have added the `erlfmt-ignore` comment to address this issue.  Here follows the new explanation from the readme:

We found that mostly it is possible to format erlang code in an at least somewhat acceptable way, but exceptions do occur.
We have introduced the `erlfmt-ignore` comment, which when placed before a top-level expression, will indicate to `erlfmt` to skip over that expression, leave it as is and move on to the next expression.

```erlang formatted matrix
%% erlfmt-ignore
-define(DELTA_MATRIX, [
    [0,   0,   0,   0,   0,   0],
    [0, -16,   0,   0,   0,   0],
    [0,   0,  15,   0,   0,   0],
    [0,   0,   0,   6,   0,   0],
    [0, -16,   0,   0, -14,   0],
    [0,   0,  15,   0,   0,   0]
]).
```

**Only top-level expression are supported.**
Nested expressions are not supported, for example expressions inside functions.
